cvmfs/download.cc

```diff
@@ -1955,6 +1961,87 @@ void DownloadManager::ProbeHosts() {
 }
 
 
+bool DownloadManager::GeoSortServers(std::vector<std::string> *servers,
+                    std::vector<uint64_t> *output_order) {
+  if (!servers) {return false;}
+  if (servers->size() == 1) {
+    if (output_order) {
+      output_order->clear();
+      output_order->push_back(0);
+    }
+    return true;
+  }
+
+  std::vector<std::string> host_chain;
+  GetHostInfo(&host_chain, NULL, NULL);
+
+  std::vector<std::string> server_dns_names;
+  server_dns_names.reserve(servers->size());
+  for (unsigned i = 0; i < servers->size(); ++i) {
+    std::string host = dns::ExtractHost((*servers)[i]);
+    server_dns_names.push_back(host.empty() ? (*servers)[i] : host);
+  }
+  std::string host_list = JoinStrings(server_dns_names, ",");
+
+  // Protect against concurrent access to prng_
+  pthread_mutex_lock(lock_options_);
+  // Determine random hosts for the Geo-API query
+  vector<string> host_chain_shuffled = Shuffle(host_chain, &prng_);
+  pthread_mutex_unlock(lock_options_);
+
+  // Request ordered list via Geo-API
+  bool success = false;
+  unsigned max_attempts = std::min(servers->size(), size_t(3));
+  vector<uint64_t> geo_order(servers->size());
+  for (unsigned i = 0; i < max_attempts; ++i) {
+    string url = host_chain_shuffled[i] + "/api/v1.0/geo/@proxy@/" + host_list;
+    LogCvmfs(kLogDownload, kLogDebug,
+             "requesting ordered server list from %s", url.c_str());
+    JobInfo info(&url, false, false, NULL);
+    Failures result = Fetch(&info);
+    if (result == kFailOk) {
+      string order(info.destination_mem.data, info.destination_mem.size);
+      free(info.destination_mem.data);
+      bool retval = ValidateGeoReply(order, servers->size(), &geo_order);
+      if (!retval) {
+        LogCvmfs(kLogDownload, kLogDebug | kLogSyslogWarn,
+                 "retrieved invalid GeoAPI reply from %s [%s]",
+                 url.c_str(), order.c_str());
+      } else {
+        LogCvmfs(kLogDownload, kLogDebug | kLogSyslog,
+                 "geographic order of servers retrieved from %s",
+                 dns::ExtractHost(host_chain_shuffled[i]).c_str());
+        LogCvmfs(kLogDownload, kLogDebug, "order is %s", order.c_str());
+        success = true;
+        break;
+      }
+    } else {
+      LogCvmfs(kLogDownload, kLogDebug | kLogSyslogWarn,
+               "GeoAPI request %s failed with error %d [%s]",
+               url.c_str(), result, Code2Ascii(result));
+    }
+  }
+  if (!success) {
+    LogCvmfs(kLogDownload, kLogDebug | kLogSyslogWarn,
```

Comment:
There is a duplicate of this log in `ProbeGeo()`, which I think should be removed.